Zend/zend_objects.c

```diff
@@ -192,21 +192,30 @@ ZEND_API zend_object* ZEND_FASTCALL zend_objects_new(zend_class_entry *ce)
 
 ZEND_API void ZEND_FASTCALL zend_objects_clone_members(zend_object *new_object, zend_object *old_object)
 {
+	zval *src, *dst, *end, *slot;
+	zend_property_info *prop_info;
+
 	if (old_object->ce->default_properties_count) {
-		zval *src = old_object->properties_table;
-		zval *dst = new_object->properties_table;
-		zval *end = src + old_object->ce->default_properties_count;
+		src = old_object->properties_table;
+		dst = new_object->properties_table;
+		end = src + old_object->ce->default_properties_count;
 
 		do {
 			i_zval_ptr_dtor(dst);
 			ZVAL_COPY_VALUE_PROP(dst, src);
 			zval_add_ref(dst);
 			if (UNEXPECTED(Z_ISREF_P(dst)) &&
 					(ZEND_DEBUG || ZEND_REF_HAS_TYPE_SOURCES(Z_REF_P(dst)))) {
-				zend_property_info *prop_info = zend_get_property_info_for_slot(new_object, dst);
+				prop_info = zend_get_property_info_for_slot(new_object, dst);
 				if (ZEND_TYPE_IS_SET(prop_info->type)) {
 					ZEND_REF_ADD_TYPE_SOURCE(Z_REF_P(dst), prop_info);
 				}
+			} else if (UNEXPECTED(old_object->ce->clone)) {
+				prop_info = zend_get_property_info_for_slot(new_object, dst);
+				if (UNEXPECTED(prop_info->flags & ZEND_ACC_READONLY)) {
+					slot = OBJ_PROP(new_object, prop_info->offset);
+					Z_PROP_FLAG_P(slot) = IS_PROP_UNINIT;
```

Comment:
```
php > class Foo { public readonly int $x = 0; }
Fatal error: Readonly property Foo::$x cannot have default value in php shell code on line 1
```

I think the test failures might be related to the fact php internally expects properties with defined values not to have the type IS_PROP_UNINIT, but I'm not completely certain. UNINIT means the value would be IS_UNDEF. In the failing clone tests, they'd have type `int` for `$this->foo = 0`, so it sees the property's type is defined and emits the error.

I also don't remember how to properly solve that or what other edge cases there may be

Reply:
Thanks for chiming in @TysonAndre.
I've figured our that there is another field in the zval that can be used as a flag for what we need: `u1.v.u.extra`, which looks unused.
I've updated the PR to used it.
I've also figure out other places where this flag should be checked.
There's a segfault on `tests/classes/clone_003.phpt` which I don't understand. Anyone would for me please?